The Big Query

ahandley opened the pull request:
Query functionality + tests

quoctruong commented on Google.PowerShell.IntegrationTests/BigQuery/BigQuery.BqJob.Tests.ps1:
```diff
@@ -63,4 +63,53 @@ Describe "Get-BqJob" {
+
+    It "should use legacy SQL when asked" {
+        $job = Start-BqJob -Query "select * from $datasetName.table_$r where Year > 1900" -UseLegacySql
+        $job.Configuration.Query.UseLegacySql | Should Be True
```
We should check the result of the job too.

ahandley replied:
Should have had a TODO in for that one too.  I need Receive- for that, and I'll add it in asap.

quoctruong commented on Google.PowerShell/BigQuery/BqJob.cs:
```diff
@@ -178,5 +178,257 @@ public void DoGetRequest()
+    /// <example>
+    ///   <code>
+    /// PS C:\> $job = Start-BqJob -Query "select * from classics where Year > 1900" `
+    /// -DefaultDataset $dataset -DestinationTable $table
```
nit: -DefaultDataset should be at the same indentation level as -Query.

quoctruong commented on Google.PowerShell/BigQuery/BqJob.cs:
```diff
@@ -178,5 +178,257 @@ public void DoGetRequest()
+
+        /// <summary>
+        /// <para type="description">
+        /// The project to look for jobs in. If not set via PowerShell parameter processing, it will
```
The comment seems to be for Get-BqJob.

ahandley replied:
I thought it worked, but I re-worded it to be more exact.

quoctruong commented on Google.PowerShell/BigQuery/BqJob.cs:
```diff
@@ -178,5 +178,257 @@ public void DoGetRequest()
+        /// </para>
+        /// </summary>
+        [Parameter(Mandatory = false)]
+        [Alias("Synch")]
```
I think the alias should be "Synchronous" instead of just "Synch". That way, user can still use -Synch if they want.

ahandley replied:
What does -Synch do?  I can't find any docs on it.  
I went ahead and updated it anyway, though.

quoctruong replied:
I mean in PowerShell if your parameter is -FooName and on the command line, you just type -Foo then PowerShell can still resolve it to -FooName.

ahandley replied:
Oh, gotcha.  That works.

quoctruong commented on Google.PowerShell/BigQuery/BqJob.cs:
```diff
@@ -178,5 +178,257 @@ public void DoGetRequest()
+        /// </para>
+        /// </summary>
+        [Parameter(Mandatory = false, ParameterSetName = ParameterSetNames.DoQuery)]
+        public SwitchParameter DryRun { get; set; }
```
We should probably make use of the built-in -WhatIf parameter of PowerShell (when you do SupportsShouldProcess) and remove this parameter.

ahandley replied:
That'll work for now.  Lemme patch that in.

quoctruong commented on Google.PowerShell/BigQuery/BqJob.cs:
```diff
@@ -178,5 +178,257 @@ public void DoGetRequest()
+        [PropertyByTypeTransformation(TypeToTransform = typeof(Dataset),
+            Property = nameof(Dataset.DatasetReference))]
+        [Alias("Dataset")]
+        public DatasetReference DefaultDataset { get; set; }
```
I think this parameter should just be Dataset. From the user's perspective, the Default prefix does not add any value and may be confusing.

ahandley replied:
That's probably fair.  The counter argument would be that people might think all referenced tables would be looked for in that dataset, so leaving the "default" in there would help.  
Is it sufficient to explain how the default works in the parameter docs and leave it at that?

ahandley replied:
@quoctruong - bump?

ILMTitan replied:
I am very split on this, but I think I am siding with @ahandley. I think this parameter is different enough from other Dataset parameters that giving it a special name is warranted.

quoctruong replied:
I think DefaultDataset is okay as long as we document it well.

ahandley replied:
I added in a few clarifying words to the parameter description, and added a section in the cmdlet docs.  Will push that in the next commit.

ahandley commented:
Addressed all the above issues (with the exception of the Dataset thing.  Discussion open on that).  Added Shouldprocess to the query subset, and I'll continue to add that for each other section as they get coded.

ILMTitan commented on Google.PowerShell/BigQuery/BqJob.cs:
```diff
@@ -178,5 +178,257 @@ public void DoGetRequest()
+
+        protected override void ProcessRecord()
+        {
+            var Client = BigQueryClient.Create(Project);
```
In our Apiary based cmdlets put the Client in an Inherited property. I suggest you do the same here (possibly making a separate abstract subclass).

ahandley replied:
So I looked into putting it in the pre-built BqCmdlet class with the Apis.BQ Service:

```
        public BigqueryService Service { get; private set; }
        public BigQueryClient Client { get; private set; }

        public BqCmdlet()
        {
            Service = new BigqueryService(GetBaseClientServiceInitializer());
            Client = BigQueryClient.Create();  //takes a ProjectId as an argument though.
        }
```

The issue with this is that you need to initialize the BigQueryClient with a project ID.  This is not known, in most cases, until parameter sets are parsed, which is after the super calls happen.  

If I'm reading your comment the wrong way, let me know, but I think the call is currently best placed where it is.

ahandley replied:
So I tried playing around with the Lazy<>, but it still had the chicken-or-the-egg problem.  I needed to put the lazy declaration on the property, but "Project" wasn't available until the constructor.  

It's definitely messier this way, but I couldn't find a clean solution (or any at all) after messing around with the Lazy's

ILMTitan replied:
You don't have to set the readonly Lazy field in an initialize. You can set the readonly field in the constructor. From there you can reference (but not resolve) Project. Then, when Client is first resolved, Project will have a value, and everything should work.

ILMTitan commented on Google.PowerShell/BigQuery/BqJob.cs:
```diff
@@ -178,5 +178,257 @@ public void DoGetRequest()
+                    bqr.PollUntilCompleted();
+                }
+
+                WriteObject(bqr.Resource);
```
These functions should return the Resource to ProcessRecord. The WriteObject call should be done there.

ILMTitan replied:
These functions should return the Resource to `ProcessRecord`. The `WriteObject` call should be done there.

ahandley replied:
Got it.

ILMTitan commented on Google.PowerShell/BigQuery/BqJob.cs:
```diff
@@ -178,5 +178,257 @@ public void DoGetRequest()
+                        new Exception("You must select a valid BQ Job type."),
+                        "Type Not Found", ErrorCategory.ObjectNotFound, this));
+                    break;
+            }
```
You can set a single variable from all of the functions above, and do a single `WriteObject` call here (if they return the same type of resource).

ahandley replied:
Yep.  It's all a Apis.Bigquery.v2.Data.Job.  Simple enough.

ahandley commented:
Alright.  Try that.  Much better.

ILMTitan commented on Google.PowerShell/BigQuery/BqCmdlet.cs:
```diff
@@ -11,11 +13,17 @@ namespace Google.PowerShell.BigQuery
         public BqCmdlet()
         {
-            Service = new BigqueryService(GetBaseClientServiceInitializer());
+            //Service = new BigqueryService(GetBaseClientServiceInitializer());
```
Remove commented code.

ILMTitan commented on Google.PowerShell/BigQuery/BqJob.cs:
```diff
@@ -178,5 +178,251 @@ public void DoGetRequest()
+            switch (ParameterSetName)
+            {
+                case ParameterSetNames.DoQuery:
+                    result = DoQuery(Client);
```
Part of the point of doing the Client property was you didn't have to pass this parameter anymore.

quoctruong commented on Google.PowerShell/BigQuery/BqCmdlet.cs:
```diff
@@ -11,11 +13,17 @@ namespace Google.PowerShell.BigQuery
         {
-            Service = new BigqueryService(GetBaseClientServiceInitializer());
+            //Service = new BigqueryService(GetBaseClientServiceInitializer());
+            _service = new Lazy<BigqueryService>(() => new BigqueryService(GetBaseClientServiceInitializer()));
```
Not sure if you need a Lazy for the service since it's almost always used by the cmdlet.

ahandley replied:
I figured that I would use it because of consistency and because a few cmdlets don't use it, so might as well save time.

ILMTitan replied:
The ones that use Client don't use Service. I think it is fine.

ahandley commented:
Thanks!
